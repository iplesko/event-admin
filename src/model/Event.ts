export interface Event {
  id?: number,
  from: Date,
  to: Date,
  name: string,
  description: string,
}
