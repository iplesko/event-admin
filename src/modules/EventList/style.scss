.event-list {
  .card {
    margin-top: 1.6rem;
  }

  .card-title {
    font-weight: bold;
  }

  .card-body {
    button {
      margin-right: 1rem;
    }
  }

  .past-event {
    opacity: .5;
  }
}
