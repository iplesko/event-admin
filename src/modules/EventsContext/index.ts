export { default as EventsContextProvider } from './EventsContextProvider';
export { default as useEventsContext } from './hooks/useEventsContext';
