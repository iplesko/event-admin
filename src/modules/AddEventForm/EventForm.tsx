import React, { ChangeEvent, FormEvent, ReactElement, useEffect, useState } from 'react';
import { Button, FormGroup, Input, Label, Modal, ModalBody, ModalFooter, ModalHeader } from 'reactstrap';
import { Event } from '../../model/Event'
import formatDate from "../../function/formatDate";

interface Props {
  heading: string;
  buttonText: string;
  onConfirm: (e: Event) => void;
  visible: boolean;
  toggleVisibility: () => void;
  displayedEvent?: Event;
}

const empty = () => ({
  name: '', from: new Date(), to: new Date(), description: '',
});

export default ({ heading, buttonText, onConfirm, visible, toggleVisibility, displayedEvent }: Props): ReactElement => {
  const [ event, setEvent ] = useState<Event>({
    name: '', from: new Date(), to: new Date(), description: '',
  });

  const handleChangeText = (e: ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.target;
    setEvent(prevState => ({ ...prevState, [id]: value }));
  }

  const handleChangeDate = (e: ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.target;
    setEvent(prevState => ({ ...prevState, [id]: new Date(value) }));
  }

  const confirm = (e: FormEvent) => {
    onConfirm(event);
    toggleVisibility();
    e.preventDefault();
  }

  useEffect(() => {
    setEvent(displayedEvent || empty());
  }, [ visible ]);

  return (
      <div>
        <Modal isOpen={visible} toggle={toggleVisibility}>
          <form onSubmit={confirm}>
            <ModalHeader toggle={toggleVisibility}>{heading}</ModalHeader>
            <ModalBody>

              <FormGroup required>
                <Label for="name">Event name *</Label>
                <Input required type="text" id="name" value={event.name} onChange={handleChangeText}/>
              </FormGroup>

              <FormGroup required>
                <Label for="from">From *</Label>
                <Input required type="date" id="from" value={formatDate(event.from)} onChange={handleChangeDate}/>
              </FormGroup>

              <FormGroup required>
                <Label for="to">To *</Label>
                <Input required type="date" id="to" value={formatDate(event.to)} onChange={handleChangeDate}/>
              </FormGroup>

              <FormGroup>
                <Label for="description">Description</Label>
                <Input type="textarea" id="description" value={event.description} onChange={handleChangeText}/>
              </FormGroup>

            </ModalBody>
            <ModalFooter>
              <Button color="primary" type="submit">{buttonText}</Button>{' '}
              <Button color="secondary" onClick={toggleVisibility}>Cancel</Button>
            </ModalFooter>
          </form>
        </Modal>
      </div>
  );
}
