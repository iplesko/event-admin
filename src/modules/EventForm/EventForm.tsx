import React, {
  ChangeEvent, FormEvent, ReactElement, useEffect, useState,
} from 'react';
import {
  Button, FormGroup, Input, Label, Modal, ModalBody, ModalFooter, ModalHeader,
} from 'reactstrap';
import { Event } from '../../model/Event';
import formatDate from '../../function/formatDate';

interface Props {
  heading: string;
  buttonText: string;
  onConfirm: (e: Event) => void;
  visible: boolean;
  toggleVisibility: () => void;
  displayedEvent?: Event;
}

const empty = () => ({
  name: '', from: new Date(), to: new Date(), description: '',
});

export default ({
  heading, buttonText, onConfirm, visible, toggleVisibility, displayedEvent,
}: Props): ReactElement => {
  const [event, setEvent] = useState<Event>(empty());

  const handleChangeText = (e: ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.target;
    setEvent((prevState) => ({ ...prevState, [id]: value }));
  };

  const handleChangeDateRange = (e: ChangeEvent<HTMLInputElement>) => {
    let from: Date;
    let to: Date;

    const { id, value } = e.target;
    if (id === 'to') {
      to = new Date(value);
      from = event.from > to ? to : event.from;
    } else {
      from = new Date(value);
      to = from > event.to ? from : event.to;
    }
    setEvent((prevState) => ({ ...prevState, from, to }));
  };

  const confirm = (e: FormEvent) => {
    onConfirm(event);
    toggleVisibility();
    e.preventDefault();
  };

  useEffect(() => {
    if (visible) {
      setEvent(displayedEvent || empty());
    }
  }, [visible]);

  return (
    <Modal isOpen={visible} toggle={toggleVisibility}>
      <form onSubmit={confirm}>
        <ModalHeader toggle={toggleVisibility}>{heading}</ModalHeader>
        <ModalBody>

          <FormGroup required>
            <Label for="name">Event name *</Label>
            <Input required type="text" id="name" value={event.name} onChange={handleChangeText} />
          </FormGroup>

          <FormGroup required>
            <Label for="from">From *</Label>
            <Input required type="date" id="from" value={formatDate(event.from)} onChange={handleChangeDateRange} />
          </FormGroup>

          <FormGroup required>
            <Label for="to">To *</Label>
            <Input required type="date" id="to" value={formatDate(event.to)} onChange={handleChangeDateRange} />
          </FormGroup>

          <FormGroup>
            <Label for="description">Description</Label>
            <Input type="textarea" id="description" value={event.description} onChange={handleChangeText} />
          </FormGroup>

        </ModalBody>
        <ModalFooter>
          <Button color="primary" type="submit">{buttonText}</Button>
          {' '}
          <Button color="secondary" onClick={toggleVisibility}>Cancel</Button>
        </ModalFooter>
      </form>
    </Modal>
  );
};
