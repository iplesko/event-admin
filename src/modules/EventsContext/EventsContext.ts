import { createContext } from 'react';
import { Event } from '../../model/Event'

export interface EventsContextType {
  events: Event[],
  addEvent: (e: Event) => void;
  removeEvent: (eventId: number) => void;
  updateEvent: (eventId: number, e: Event) => void
}

export default createContext<EventsContextType>(undefined);
