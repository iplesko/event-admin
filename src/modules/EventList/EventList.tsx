import React, { useState } from 'react';
import './style.scss';
import {
  Badge, Button, Card, CardBody, CardSubtitle, CardText, CardTitle, Col, Row,
} from 'reactstrap';
import { useEventsContext } from '../EventsContext';
import { Event } from '../../model/Event';
import EventForm from '../EventForm';
import formatDate from '../../function/formatDate';

export default (): React.ReactElement => {
  const {
    events, updateEvent, removeEvent, addEvent,
  } = useEventsContext();

  const [addFormVisible, setAddFormVisible] = useState<boolean>(false);
  const toggleAddFormVisible = () => setAddFormVisible(!addFormVisible);

  const [editEvent, setEditEvent] = useState<Event>(null);
  const toggleEditFormVisible = () => setEditEvent(null);

  return (
    <>
      <EventForm
        heading="Add event"
        buttonText="Add"
        onConfirm={addEvent}
        visible={addFormVisible}
        toggleVisibility={toggleAddFormVisible}
      />
      <EventForm
        heading="Edit event"
        buttonText="Save"
        onConfirm={updateEvent}
        visible={editEvent !== null}
        toggleVisibility={toggleEditFormVisible}
        displayedEvent={editEvent}
      />

      <Row>
        <Col>
          <Button onClick={toggleAddFormVisible} color="success">Add event</Button>
        </Col>
      </Row>

      <Row className="event-list">
        {events.map((e) => {
          const now = new Date(Date.now());
          const past = e.from < new Date(now.setDate(now.getDate() - 1));
          return (
            <Col sm={4} key={e.id}>
              <Card className={past ? 'past-event' : ''}>
                <CardBody>
                  <CardTitle>
                    {`${e.name} `}
                    {past && <Badge color="info">Past event</Badge>}
                  </CardTitle>
                  <CardSubtitle className="text-muted">
                    {`${formatDate(e.from)} - ${formatDate(e.to)}`}
                  </CardSubtitle>
                  <CardText>{e.description}</CardText>
                  <Button color="primary" onClick={() => setEditEvent(e)}>Edit</Button>
                  <Button color="danger" onClick={() => removeEvent(e.id)}>Remove</Button>
                </CardBody>
              </Card>
            </Col>
          );
        })}
      </Row>
    </>
  );
};
