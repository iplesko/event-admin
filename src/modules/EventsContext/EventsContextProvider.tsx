import React, { ReactElement, useState } from 'react';
import EventsContext, { EventsContextType } from "./EventsContext";
import { Event } from '../../model/Event';
import useCounter from "./hooks/useCounter";

interface Props {
  children: ReactElement;
}

const compareEvents = (a: Event, b: Event): number => {
  if (a.from === b.from) {
    return 0;
  }

  return a.from > b.from ? -1 : 1;
}

export default ({ children }: Props): ReactElement => {
  const [ events, setEvents ] = useState<Event[]>([])
  const { getAndIncrement } = useCounter();

  // TODO: make this better readable
  const context: EventsContextType = {
    events: events,
    addEvent: (e: Event) => setEvents(previousEvents => ([
      ...previousEvents, { ...e, id: getAndIncrement() }
    ].sort(compareEvents))),
    removeEvent: (eventId: number) => setEvents(previousEvents => previousEvents.filter(pe => pe.id !== eventId)),
    updateEvent: (e: Event) => setEvents(previousEvents => [ ...previousEvents ].map(
        pe => pe.id === e.id ? e : pe
    ).sort(compareEvents))
  }

  return (
      <EventsContext.Provider value={context}>
        {children}
      </EventsContext.Provider>
  );
}
