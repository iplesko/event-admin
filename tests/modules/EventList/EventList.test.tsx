import React from 'react';
import { shallow } from 'enzyme';
import { Button, Card } from 'reactstrap';
import EventList from '../../../src/modules/EventList/EventList';
import { useEventsContext } from '../../../src/modules/EventsContext';
import EventForm from '../../../src/modules/EventForm';
import { Event } from '../../../src/model/Event';

jest.mock('../../../src/modules/EventsContext');

describe('EventList', () => {
  let addEventMock: jest.Mock;
  let updateEventMock: jest.Mock;
  let removeEventMock: jest.Mock;
  let testEvent: Event;

  beforeEach(() => {
    testEvent = {
      id: 123,
      name: 'test event name',
      from: new Date(12345678),
      to: new Date(87654321),
      description: 'test event description',
    };
    addEventMock = jest.fn();
    updateEventMock = jest.fn();
    removeEventMock = jest.fn();

    (useEventsContext as jest.Mock).mockReturnValue({
      events: [testEvent],
      addEvent: addEventMock,
      removeEvent: removeEventMock,
      updateEvent: updateEventMock,
    });
  });

  it('should render correctly', () => {
    const wrapper = shallow(<EventList />);

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should add new event', () => {
    const wrapper = shallow(<EventList />);

    wrapper.find(EventForm).at(0).prop('onConfirm')(testEvent);

    expect(addEventMock).toHaveBeenCalledWith(testEvent);
  });

  it('should edit event', () => {
    const wrapper = shallow(<EventList />);

    wrapper.find(EventForm).at(1).prop('onConfirm')(testEvent);

    expect(updateEventMock).toHaveBeenCalledWith(testEvent);
  });

  it('should delete event', () => {
    const wrapper = shallow(<EventList />);

    wrapper.find(Button)
      .filterWhere((button) => button.prop('color') === 'danger')
      .simulate('click');

    expect(removeEventMock).toHaveBeenCalledWith(123);
  });

  it('should render future and past event', () => {
    Date.now = jest.fn().mockReturnValue(new Date('2020-07-21').getTime());

    (useEventsContext as jest.Mock).mockReturnValue({
      events: [{
        id: 1,
        name: 'future event',
        from: new Date('2020-07-21'),
        to: new Date('2020-07-21'),
      }, {
        id: 2,
        name: 'past event',
        from: new Date('2020-07-19'),
        to: new Date('2020-07-19'),
      }],
      addEvent: addEventMock,
      removeEvent: removeEventMock,
      updateEvent: updateEventMock,
    });

    const wrapper = shallow(<EventList />);

    const cards = wrapper.find(Card);
    expect(cards.at(0).prop('className')).not.toContain('past-event');
    expect(cards.at(1).prop('className')).toContain('past-event');
  });

  it('should display add event modal', () => {
    const wrapper = shallow(<EventList />);

    wrapper.find(Button)
      .filterWhere((button) => button.prop('color') === 'success')
      .simulate('click');

    expect(wrapper.find(EventForm).at(0).prop('visible')).toBeTruthy();
  });

  it('should display edit event modal', () => {
    const wrapper = shallow(<EventList />);

    wrapper.find(Card)
      .at(0)
      .find('Button')
      .filterWhere((button) => button.prop('color') === 'primary')
      .simulate('click');

    expect(wrapper.find(EventForm).at(1).prop('visible')).toBeTruthy();
  });
});
