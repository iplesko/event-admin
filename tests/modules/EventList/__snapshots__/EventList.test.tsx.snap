// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventList should render correctly 1`] = `
"<Fragment>
  <Component heading=\\"Add event\\" buttonText=\\"Add\\" onConfirm={[Function: mockConstructor]} visible={false} toggleVisibility={[Function: toggleAddFormVisible]} />
  <Component heading=\\"Edit event\\" buttonText=\\"Save\\" onConfirm={[Function: mockConstructor]} visible={false} toggleVisibility={[Function: toggleEditFormVisible]} displayedEvent={{...}} />
  <Row tag=\\"div\\" widths={{...}}>
    <Col tag=\\"div\\" widths={{...}}>
      <Button onClick={[Function: toggleAddFormVisible]} color=\\"success\\" tag=\\"button\\">
        Add event
      </Button>
    </Col>
  </Row>
  <Row className=\\"event-list\\" tag=\\"div\\" widths={{...}}>
    <Col sm={4} tag=\\"div\\" widths={{...}}>
      <Card className=\\"past-event\\" tag=\\"div\\">
        <CardBody tag=\\"div\\">
          <CardTitle tag=\\"div\\">
            test event name 
            <Badge color=\\"info\\" pill={false} tag=\\"span\\">
              Past event
            </Badge>
          </CardTitle>
          <CardSubtitle className=\\"text-muted\\" tag=\\"div\\">
            1970-01-01 - 1970-01-02
          </CardSubtitle>
          <CardText tag=\\"p\\">
            test event description
          </CardText>
          <Button color=\\"primary\\" onClick={[Function: onClick]} tag=\\"button\\">
            Edit
          </Button>
          <Button color=\\"danger\\" onClick={[Function: onClick]} tag=\\"button\\">
            Remove
          </Button>
        </CardBody>
      </Card>
    </Col>
  </Row>
</Fragment>"
`;
